### PR TITLE
feat(swapper): add approveAmount swapper API method

### PR DESCRIPTION
**Pull request by 0xApotheosis**
Draft PR for @NeOMakinG to take over and finish https://github.com/shapeshift/web/issues/2263

Web implementation: https://github.com/shapeshift/web/pull/2419

### CODE REVIEW

File: packages/swapper/src/api.ts [hunk 1]
```diff
@@ -189,7 +194,7 @@ export enum SwapErrorTypes {
 }
 
 export interface Swapper<T extends ChainId> {
-  /** Human readable swapper name */
+  /** Human-readable swapper name */
```
Comment (gomesalexandre) [inline packages/swapper/src/api.ts:197]: ❤️

File: packages/swapper/src/api.ts [hunk 2]
```diff
@@ -129,6 +129,10 @@ export type ApproveInfiniteInput<C extends ChainId> = {
   wallet: HDWallet
 }
 
+export type ApproveAmountInput<C extends ChainId> = ApproveInfiniteInput<C> & {
```
Comment (gomesalexandre) [inline packages/swapper/src/api.ts:132]: Rename `ApproveInfiniteInput` to `ApproveInput` and make `ApproveInfiniteInput` an alias of the latter?

File: packages/swapper/src/swappers/cow/CowSwapper.ts [hunk 3]
```diff
@@ -76,6 +78,12 @@ export class CowSwapper implements Swapper<KnownChainIds.EthereumMainnet> {
     return cowApproveInfinite(this.deps, args)
   }
 
+  async approveAmount(): Promise<string> {
+    throw new SwapError('CowSwapper: approveAmount unimplemented', {
+      code: SwapErrorTypes.RESPONSE_ERROR
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/cow/CowSwapper.ts:83]: Not sure that's the best error code semantically, but I don't believe adding a code for unimplemented methods is a good solution either so let's go with this!

File: packages/swapper/src/swappers/cow/CowSwapper.ts [hunk 4]
```diff
@@ -76,6 +78,12 @@ export class CowSwapper implements Swapper<KnownChainIds.EthereumMainnet> {
     return cowApproveInfinite(this.deps, args)
   }
 
+  async approveAmount(): Promise<string> {
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/cow/CowSwapper.ts:81]: - This never returns a string on the unimplemented implementations.
- The implemented ones are also not *guaranteed* to return a string, they can throw and thus return `void`.
Let's change the type definition of the implemented ones to `string | void`, and the one from the unimplemented ones to `void`?

File: packages/swapper/src/swappers/zrx/zrxApproveInfinite/zrxApproveInfinite.ts [hunk 5]
```diff
@@ -1,37 +1,67 @@
...
+  return await grantAllowance<T>({
+    quote: approvalQuote,
+    wallet,
+    adapter,
+    erc20Abi,
+    web3
+  })
+}
+
+export async function zrxApproveAmount<T extends EvmSupportedChainIds>(
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/zrx/zrxApproveInfinite/zrxApproveInfinite.ts:41]: This module is named `zrxApproveInfinite`, it should either be renamed, or this should be in a whole new module (preference for the former vs. having to search through many different modules, this should be colocated IMO)

**Review by gomesalexandre — COMMENTED** [review]
A few initial comments but overall looks good

File: packages/swapper/src/swappers/cow/CowSwapper.ts (hunk 4 above)
Comment (kaladinlight) [reply packages/swapper/src/swappers/cow/CowSwapper.ts:81]: More accurately, this function would be assigned a return type of never (which does satisfy the swapper interface), though we don't really interact with specific swapper types in the client.

Considering this is an async function that should be wrapped in a try/catch, I don't see too big of a problem with this current implementation and it follows how we have handled unimplemented function up until this point as far as I am aware.

File: packages/swapper/src/swappers/zrx/zrxApproveInfinite/zrxApproveInfinite.ts (hunk 5 above)
Comment (kaladinlight) [reply packages/swapper/src/swappers/zrx/zrxApproveInfinite/zrxApproveInfinite.ts:41]: or even better, wax all of these nested directories and just have the swapper class with the logic contained within there :) not a fan of this bloated setup personally. obv not for this PR.

I would go with rename as well if we want to change at this time

**Review by gomesalexandre — COMMENTED** [review]
Tested locally while linked in the matching web PR and this does what it says on the box.

File: packages/swapper/src/swappers/zrx/zrxApproveInfinite/zrxApproveInfinite.ts (hunk 5 above)
Comment (NeOMakinG) [reply packages/swapper/src/swappers/zrx/zrxApproveInfinite/zrxApproveInfinite.ts:41]: Agree, and so how would we call the module containing these 2 functions? `srxApprove`?

Comment (gomesalexandre) [reply packages/swapper/src/swappers/zrx/zrxApproveInfinite/zrxApproveInfinite.ts]: Precisely @NeOMakinG, that's what I had in mind

**Review by gomesalexandre — APPROVED** [review]
Looks good now, comments have been tackled.

Gave it a sanity check while linked to the web PR and still works as expected.

Note that until implemented, Thor/CoW exact allowances in the trade component will fail
Given:
1. This wasn't in the spec
2. Those two are under feature flag

I'm fine with getting this one (and web) in, and tackling Thor and CoW as follow-ups.

Well done ser 🍬

**Pull request merged**